lightly/loss/dcl_loss.py

```diff
@@ -0,0 +1,219 @@
+            weight_fn is None.
+        gather_distributed:
+            If True then negatives from all gpus are gathered before the 
+            loss calculation.
+    
+    Examples:
+
+        >>> loss_fn = DCLLoss(temperature=0.07)
+        >>>
+        >>> # generate two random transforms of images
+        >>> t0 = transforms(images)
+        >>> t1 = transforms(images)
+        >>>
+        >>> # embed images using some model, for example SimCLR
+        >>> out0 = model(t0)
+        >>> out1 = model(t1)
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+        
+    """
+    def __init__(
+        self,
+        temperature: float = 0.1,
+        weight_fn: Optional[Callable[[Tensor, Tensor], Tensor]] = None,
+        gather_distributed: bool = False,
+    ):
+        super().__init__()
+        self.temperature = temperature
+        self.weight_fn = weight_fn
+        self.gather_distributed = gather_distributed
+
+    def forward(
+        self,
+        out0: Tensor,
+        out1: Tensor,
+    ) -> Tensor:
+        """Forward pass of the DCL loss.
+        
+        Args:
+            out0:
+                Output projections of the first set of transformed images.
+                Shape: (batch_size, embedding_size)
+            out1:
+                Output projections of the second set of transformed images.
+                Shape: (batch_size, embedding_size)
+        
+        Returns:
+            Mean loss over the mini-batch.
+        """
+        # normalize the output to length 1
+        out0 = out0_all = torch.nn.functional.normalize(out0, dim=1)
+        out1 = out1_all = torch.nn.functional.normalize(out1, dim=1)
+
+        if self.gather_distributed and dist.world_size() > 1:
+            # gather representations from other processes if necessary
+            out0_all = torch.cat(dist.gather(out0), 0)
+            out1_all = torch.cat(dist.gather(out1), 0)
+
+        # calculate symmetric loss
+        loss0 = self._loss(out0, out1, out0_all, out1_all)
+        loss1 = self._loss(out1, out0, out1_all, out0_all)
+        return 0.5 * (loss0 + loss1)
+
+    def _loss(self, out0, out1, out0_all, out1_all):
+        """Calculates DCL loss for out0 with respect to its positives in out1
+        and the negatives in out1, out0_all, and out1_all.
+        
+        This code implements Equation 6 in [0], including the sum over all images `i`
+        but with `k` fixed at 0.
+        
+        Args:
+            out0:
+                Output projections of the first set of transformed images.
+                Shape: (batch_size, embedding_size)
+            out1:
+                Output projections of the second set of transformed images.
+                Shape: (batch_size, embedding_size)
+            out0_all:
+                Output projections of the first set of transformed images from
+                all processes.
```

Comment:
Add/ mention that "all processes" is for distributed processing (e.g. from all GPUs)